### PR TITLE
[Feat] DeepSpeed single file saving

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

DeepSpeed has released a new version which contains all gathering weights to a single process and saving. As discussed in #6691 it's probably best for now to prioritise one file saving as this is easiest to deploy etc. As a result I've made this the default behaviour when using ZeRO 3.

In addition, if a user uses the `configure_sharded_model` hook, it is necessary to call this in the `load_from_checkpoint` function as the test indicates via the model hook. This could potentially be done for the user, however right now I'll make it clear the docs in a separate PR that this needs to be handled manually.

cc @ananthsub @tchaton 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -472,17 +480,26 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
+            checkpoint: The checkpoint state dictionary
             filepath: write-target file's path
-            weights_only: saving model weights only
         """
         if self.world_size > 1 and self.zero_stage_3:
+            save_dir = self._filepath_to_dir(filepath)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:487]: Can be moved next to `_exclude_keys`.

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 2]
```diff
@@ -472,17 +480,26 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
...
+            if self.save_full_weights:
+                # todo: expose this as a unprotected function in deepspeed
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:489]:
```suggestion
                # todo: expose this as an unprotected function in deepspeed
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -472,17 +480,26 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
...
+                state_dict = self.deepspeed_engine._zero3_consolidated_fp16_state_dict()
+                if self.is_global_zero:
+                    # Remove module portion of the checkpoint before saving
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:492]: Can you explain what is a module portion in the comment ?

File: tests/plugins/test_deepspeed_plugin.py [hunk 4]
```diff
@@ -28,6 +28,9 @@ def __init__(self):
     def configure_sharded_model(self) -> None:
         self.linear = torch.nn.Linear(32, 2)
 
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        self.configure_sharded_model()
```
Comment (tchaton) [inline tests/plugins/test_deepspeed_plugin.py:32]: Any way we could add this to LightningModule directly under the condition of DeepSpeed v3 ?

**Review by tchaton — APPROVED** [review]
Overall LGTM ! Small nits

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/deepspeed.py:492]: is this intending to not save the full weights on ranks other than rank 0? if so, does the check need to be if the self.global_rank > 1 instead of self.world_size > 1 ?

File: tests/plugins/test_deepspeed_plugin.py (hunk 4 above)
Comment (ananthsub) [reply tests/plugins/test_deepspeed_plugin.py:32]: Not cleanly. All LightningModule instances would need to call `super().on_load_checkpoint(checkpoint)` then, right? It'd also be strange since people could try to load from checkpoint outside of training
```
if self.trainer and isinstance(self.trainer.training_type_plugin, DeepSpeedPlugin)
```
while also needing to check if deepspeed is installed

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -472,17 +480,26 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
...
+                    state_dict = {k.partition('module.')[2]: state_dict[k] for k in state_dict.keys()}
+                    checkpoint['state_dict'] = state_dict
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/deepspeed.py:494]: n00b question: is this flipped? what does the state dict look like for non rank-0 ranks?

File: tests/plugins/test_deepspeed_plugin.py (hunk 4 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:32]: I was thinking, we could inspect the lightning module cls and see if `configure_sharded_model` has been overridden. If so then we can call this for the user, which would mean it would work.

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 5 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:494]: What do you mean flipped? It's `None` on every other rank: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L1713

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 3 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:492]: I shamelessly took the code from https://discuss.pytorch.org/t/prefix-parameter-names-in-saved-model-if-trained-by-multi-gpu/494/2

I'll clear it up in the comment, but recall we always wrap our model in a wrapper class (`LightningDeepSpeedModule`). This means the state dict includes this as a prefix across the keys. I delete this prefix before saving on the keys :)

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by kaushikb11** [comment]
Hi @SeanNaren, this particular [test](https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=5203&view=logs&j=cb44143e-1b72-550b-97cf-26ffc4f8d032&t=e7b66bef-0fa1-58f6-bb54-96c6fcb28d65&l=2269) is failing.

**Comment by SeanNaren** [comment]
I'm also noticing the container is failing when DeepSpeed is added to the extras. I might move to having DeepSpeed installed within the GPU container instead of the extras file.

**Review by ananthsub — COMMENTED** [review]
@SeanNaren totally separate from this PR, but have you seen checkpoint saving become a bottleneck at all with deepspeed? I think there's room to optimize the saving logic inside the lightning checkpoint callback, since we are potentially re-running the broadcast/state dict modification 2 times here: https://github.com/PyTorchLightning/pytorch-lightning/blob/fe0d08899eba94d275ff42253f495d9e70d86f89/pytorch_lightning/callbacks/model_checkpoint.py#L278-L286

**Comment by SeanNaren** [comment]
> pytorch-lightning/pytorch_lightning/callbacks/model_checkpoint.py

If multiple of these modes are used, then definitely it would be an expensive operation. Currently I've only tested with `save_last` to ensure the final model can be saved. There also seems to be a limit in terms of the the all gather required to save, but this might be from Lightning not de-allocating memory (need to check after this PR is merged).

```
Rudimentary numbers I collected for 8 A100s using minGPT and using time probing when saving the model to disk

13B parameters with ZeRO 3 + offloading
Time taken to consolidate 10.642701387405396s (picked one of the 8 values logged from the GPUs, all were about the same)
Time taken to dump 95.15039467811584s

20B parameters with ZeRO 3 + offloading
Blows up CUDA OOM

20B parameters with ZeRO 3 + offloading + deepspeed checkpointing
Blows up CUDA OOM
```
